File: algorithms/0844.Backspace String Compare/solution.py
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        si, ti, sb, tb = len(S) - 1, len(T) - 1, 0, 0
        while si >= 0 or ti >= 0:
            sc = S[si] if si >= 0 else ''
            tc = T[ti] if ti >= 0 else ''
            if sc == '#':
                si -= 1
                sb += 1
                continue
            if tc == '#':
                ti -= 1
                tb += 1
                continue
            if sb:
                si -= 1
                sb -= 1
                continue
            if tb:
                ti -= 1
                tb -= 1
                continue
            if sc != tc:
                return False
            si -= 1
            ti -= 1
        return True

File: algorithms/0844.Backspace String Compare/solution.go
package main

func backspaceCompare(S string, T string) bool {
	var sb, tb int
	si := len(S) - 1
	ti := len(T) - 1
	for si >= 0 || ti >= 0 {
		sc := getChar(S, si)
		tc := getChar(T, ti)
		if sc == '#' {
			si--
			sb++
			continue
		}
		if tc == '#' {
			ti--
			tb++
			continue
		}
		if sb > 0 {
			si--
			sb--
			continue
		}
		if tb > 0 {
			ti--
			tb--
			continue
		}
		if sc != tc {
			return false
		}
		si--
		ti--
	}
	return true
}

func getChar(str string, index int) byte {
	if index < 0 {
		return ' '
	}
	return str[index]
}
